electron/syntheticsGenerator.ts

```diff
@@ -0,0 +1,365 @@
+      return true;
+    } else if (action.name === 'click') {
+      return this.previousContext?.frameUrl === frameUrl && action.signals.length > 0;
+    }
+    return false;
+  }
+
+  generateStepStart(name: string) {
+    this.insideStep = true;
+    const formatter = new PlaywrightGenerator.JavaScriptFormatter(this.getDefaultFormatterOffset());
+    formatter.add(`step(${quote(name)}, async () => {`);
+    return formatter.format();
+  }
+
+  generateStepEnd() {
+    if (!this.insideStep) {
+      return '';
+    }
+    this.insideStep = false;
+    const formatter = new PlaywrightGenerator.JavaScriptFormatter(this.getDefaultFormatterOffset());
+    formatter.add(`});`);
+    return formatter.format();
+  }
+
+  generateHeader() {
+    const formatter = new PlaywrightGenerator.JavaScriptFormatter(0);
+    formatter.add(`
+      const { journey, step, expect } = require('@elastic/synthetics');
+
+      journey('Recorded journey', async ({ page, context }) => {`);
+    return formatter.format();
+  }
+
+  generateFooter() {
+    return `});`;
+  }
+
+  /**
+   * Generates JavaScript code from a custom set of steps and nested actions.
+   *
+   * This function makes no assumptions about where steps should be created,
+   * and instead follows the step definitions the caller has defined.
+   * @param steps IR to use for code generation
+   * @returns a list of the code strings outputted by the generator
+   */
+  generateFromSteps(steps: Steps): string {
+    const text = [];
+    if (this.isSuite) {
+      text.push(this.generateHeader());
+    }
+    this.varsToHoist = this.findVarsToHoist(steps);
+    text.push(this.generateHoistedVars());
+    for (const step of steps) {
+      if (step.actions.length === 0) throw Error('Cannot process an empty step');
+      const name = step.name ?? step.actions[0].title ?? actionTitle(step.actions[0].action);
+      text.push(this.generateStepStart(name ?? ''));
+
+      for (const action of step.actions) {
+        const actionText = this.generateAction(action);
+        if (actionText.length) text.push(actionText);
+      }
+
+      text.push(this.generateStepEnd());
+    }
+    if (this.isSuite) {
+      text.push(this.generateFooter());
+    }
+
+    return text.filter(s => !!s).join('\n');
+  }
+
+  generateHoistedVars() {
+    const formatter = new PlaywrightGenerator.JavaScriptFormatter(this.getDefaultFormatterOffset());
+    for (const varName of this.varsToHoist) {
+      formatter.add(`let ${varName};`);
+    }
+    return formatter.format();
+  }
+
+  isVarHoisted(varName: string) {
+    return this.varsToHoist.indexOf(varName) >= 0;
+  }
+
+  getDefaultFormatterOffset() {
+    return this.isSuite ? 2 : 0;
+  }
+
+  /**
+   * We need to hoist any page or popup alias that appears in more than one step.
+   * @param steps the step IR to evaluate
+   * @returns an array that contains the names of all variables that need to be hoisted
+   */
+  findVarsToHoist(steps: Steps): string[] {
```

Comment:
After seeing 3-level nested loops I was hoping to find a bit simpler way, but I couldn't come up with anything better so I will leave it to it. :)

Reply:
The complexity is unfortunately tied to the data structure. We could create some recursive function instead if you prefer; would reduce nesting but increase the logical acuity needed to trace the code.

Reply:
Yes, I agree that this is due to the data structure. It seems to be best to leave it like that and avoid additional effort just to avoid the nested loops. Thanks for sharing your thoughts.

Reply:
@vigneshshanmugam [suggested](https://github.com/elastic/synthetics-recorder/pull/211#discussion_r861388311) a few ways to improve it below, I've added those.